### PR TITLE
Add summary to setup action

### PR DESCRIPTION
**Pull request by jprinet**
### Description
Add capabilities to display a GitHub job summary for any Maven build

### Changes
- The `maven-setup` action is now a first-class citizen and ordered first in the `README` as addressing more use cases
- The summary options (`skip-*`) are transferred from the `maven-publish-build-scan` to the `maven-setup` action
- The `build-scan-shared` module has been refactored to reflect the following changes:
  - Each Maven build generates a build metadata file in `~/.m2/.gradle-enterprise/build-scan-metadata`
  - When a Build Scan is not published:
    - the scan dump is copied to `$RUNNER_TMP/maven-build-scan-data/build-scan-data`
    - the build metadata file is copied to `$RUNNER_TMP/maven-build-scan-data/build-scan-metadata`
  - The `maven-setup` post action collects all unpublished items and uploads them as a workflow artifact
  - When Maven is invoked to republish a Build Scan, the metadata file is not created but enriched with the Build Scan Link

### CODE REVIEW

File: README.md
Comment (runningcode) [inline README.md:8]: for my understanding, the workflow summary is available even for builds that do not publish scans?

File: README.md
Comment (runningcode) [inline README.md:40]:
```suggestion
By default, a comment will be added to the pull-request with the summary (can be skipped if `skip-pr-comment` is set to `true`):
```

I like that we're adding this comment by default. I think it is the best user experience but it does differ from how the Gradle action works.

File: README.md
Comment (runningcode) [inline README.md:84]: I think we discussed this before, but the next Develocity maven extension will not feature these values like `ALWAYS`, `ON_FAILURE`, and `ON_DEMAND`. It will only have conditional publication. `onlyIf`. Can you remind me why we didn't align this API?

File: README.md
Comment (runningcode) [inline README.md:92]: I wonder if this additional maven extension could cause build cache misses. Have we tested if it causes "goal implementation changed"?

File: build-scan-shared/__tests__/summary/dump.test.ts
Comment (runningcode) [inline build-scan-shared/__tests__/summary/dump.test.ts:37]: what is `dump` in this context? is it the build scan dump?
also is there a test for the conditional build scan publication?

File: maven-setup/action.yml
Comment (runningcode) [inline maven-setup/action.yml:32]: Is this a Github Project ID or a Develocity Project ID?

File: maven-setup/action.yml
Comment (runningcode) [inline maven-setup/action.yml:25]: can we define these variables with positives to avoid a double negative? Double negatives are confusing to reason about.
For example `write-pr-comment` with a default of `true`. Then you can set it to false to disable.
(same comment applies to all the other properties here)

File: maven-build-scan-capture-extension/src/main/java/com/gradle/MavenBuildScanCaptureListener.java
Comment (runningcode) [inline maven-build-scan-capture-extension/src/main/java/com/gradle/MavenBuildScanCaptureListener.java:63]: ah, we're going to have to come up with a backward compatibility strategy for this action.

**Review by runningcode — COMMENTED** [review]
Left some comments, overall looks good!

File: README.md
Comment (jprinet) [reply README.md:8]: Yes

File: README.md
Comment (jprinet) [reply README.md:84]: as far as I remember, I didn't want to execute arbitrary code in the Maven process. That would also force to ship a parser / interpreter.

File: build-scan-shared/__tests__/summary/dump.test.ts
Comment (jprinet) [reply build-scan-shared/__tests__/summary/dump.test.ts:37]: `dump` is the name of the function under test

Comment (jprinet) [reply build-scan-shared/__tests__/summary/dump.test.ts:37]: > also is there a test for the conditional build scan publication?

by conditional build scan publication, I'm assuming you are referring to `ON_DEMAND` / `ON_FAILURE` flags. Those flags are just used to condition unpublished build scans capture.

File: maven-setup/action.yml
Comment (jprinet) [reply maven-setup/action.yml:25]: good point, let me align with the [gradle-setup action](https://github.com/gradle/actions/tree/main/setup-gradle#build-reporting) and name it `add-X`

File: build-scan-shared/__tests__/summary/dump.test.ts
Comment (jprinet) [reply build-scan-shared/__tests__/summary/dump.test.ts:37]: I have a higher level test in the extension code

**Comment by jprinet** [comment]
> @jprinet , what is a timeline for this change? I am now updating this action to work with GE and DV plugins. That will be a pretty big change, and I'd like to base it on top of this changes.

As discussed directly, I'll take care of the migration

File: README.md
Comment (jprinet) [reply README.md:92]: There is no "goal implementation changed" with the extension applied, the cache keys are consistent.
See [this comparison](https://ge.solutions-team.gradle.com/c/dmzfighnwjyro/n4smq2zqiug5k/goal-inputs) for instance

File: maven-build-scan-capture-extension/src/main/java/com/gradle/MavenBuildScanCaptureListener.java
Comment (jprinet) [reply maven-build-scan-capture-extension/src/main/java/com/gradle/MavenBuildScanCaptureListener.java:63]: I will postpone merging this PR in order to directly integrate it with the Develocity API.
There is no need to rush a version with the GE API

File: maven-setup/action.yml
Comment (jprinet) [reply maven-setup/action.yml]: done

File: README.md
Comment (runningcode) [inline README.md:9]: If this is enabled by default, I would adjust the wording here.
```suggestion
The summary is also added as a pull-request comment in a pull-request context. This can be disabled via <link> 
```

File: README.md
Comment (runningcode) [inline README.md:21]: I know this is a bit late here, but I was just thinking, since this doesn't set up maven itself, perhaps the name is misleading. Does this add maven to the path?
I can't think of a better name, maybe `maven-summary-setup`? I wonder what @bigdaz would think here.

File: README.md
Comment (runningcode) [reply README.md:84]: got it, thanks

File: README.md
Comment (runningcode) [reply README.md:92]: Very nice!

File: README.md
Comment (runningcode) [inline README.md:116]: Ok so two thoughts here.
1. This (and the others) should say `Whether or not to enable unpublished scans capture`

2. Can we just get rid of all the `whether or not` and instead say `Enable Build Scan links capture` to make it more concise?

File: README.md
Comment (runningcode) [inline README.md:126]:
```suggestion
When submitting a pull request from a forked Github repository, a GitHub workflow that validates the change cannot publish a Build Scan® 
```

File: README.md
Comment (runningcode) [inline README.md:127]:
```suggestion
as workflows from forked repositories do not have access to secrets. An access token is required to publish a Build Scan®.
```

File: README.md
Comment (runningcode) [inline README.md:189]: What is the behavior if this list is missing? Are no scans published or are all scans published?

File: build-scan-shared/__tests__/summary/dump.test.ts
Comment (runningcode) [reply build-scan-shared/__tests__/summary/dump.test.ts:37]: got it.

File: maven-build-scan-capture-extension/src/main/java/com/gradle/MavenBuildScanCaptureListener.java
Comment (runningcode) [reply maven-build-scan-capture-extension/src/main/java/com/gradle/MavenBuildScanCaptureListener.java:63]: Sounds good.

**Review by runningcode — COMMENTED** [review]
Added another set of comments looking close! Now that Daz is back, maybe he also can take a look.

**Comment by bigdaz** [comment]
Is there still a plan to rename the repository `gradle/github-actions` to `gradle/develocity-actions`? I think the current name is not ideal.

**Comment by jprinet** [comment]
> Is there still a plan to rename the repository `gradle/github-actions` to `gradle/develocity-actions`? I think the current name is not ideal.

Yes this will happen once this PR gets merged + a following one to rely on DV API instead of GE API.

File: README.md
Comment (jprinet) [reply README.md:21]: We already discussed this in the [naming specification](https://docs.google.com/document/d/1CUt2IQIaBK6D7xihD2MT9_39Zvi3tRtuSW1dhnB7GuE/edit#heading=h.pr5426ll05no) and decided to keep it generic to allow further capabilities in the future (CI injection?)

File: README.md
Comment (jprinet) [reply README.md:116]: Thanks, will apply suggestion 2., it is way more readable 👍

File: README.md
Comment (jprinet) [reply README.md:189]: All scans would be published, I changed the comment to:
> The action will publish Build Scans® if the initial pull-request author belongs to the list or if the list is empty.

**Review by runningcode — APPROVED** [review]
This looks good from my end. Please also get a full review from @bigdaz since I'm no GH action expert🙏

**Review by bigdaz — APPROVED** [review]
LGTM

**Pull request merged**
